test/scripts/products/h22-validate-that-rate-limit-service-is-working-as-expected/h22-validate-that-rate-limit-service-is-working-as-expected.go

```diff
@@ -0,0 +1,504 @@
+
+	applicationPlanID, err := threescaleClient.CreateApplicationPlan(accessToken,
+		serviceID,
+		fmt.Sprintf("%s-api-plan", baseName),
+	)
+	if err != nil {
+		return "", err
+	}
+	fmt.Printf("  ✔️  Created Application Plan ID: %s\n", applicationPlanID)
+
+	userKey, err := threescaleClient.CreateApplication(accessToken,
+		accountID,
+		applicationPlanID,
+		fmt.Sprintf("%s-api-app", baseName),
+		fmt.Sprintf("%s-api-app", baseName),
+	)
+	if err != nil {
+		return "", err
+	}
+	fmt.Printf("  ✔️  User key: %s\n", userKey)
+
+	time.Sleep(5 * time.Second)
+
+	if err := threescaleClient.DeployProxy(accessToken, serviceID); err != nil {
+		return "", err
+	}
+	fmt.Println("  ✔️  Proxy deployed")
+
+	time.Sleep(5 * time.Second)
+
+	endpoint, err := threescaleClient.PromoteProxy(accessToken, serviceID, "sandbox", "production")
+	if err != nil {
+		return "", err
+	}
+	fmt.Printf("  ✔️  Promoted proxy. Endpoint: %s\n", endpoint)
+
+	return fmt.Sprintf("%s?user_key=%s", endpoint, userKey), nil
+}
+
+// getRedisHosts finds the host of the rate limiting Redis instance
+func getRedisHost(ctx context.Context, client k8sclient.Client, namespace string) (string, error) {
+	secret := &corev1.Secret{}
+	if err := client.Get(ctx, k8sclient.ObjectKey{
+		Name:      "ratelimit-service-redis-rhoam",
+		Namespace: namespace,
+	}, secret); err != nil {
+		return "", err
+	}
+
+	host, ok := secret.Data["uri"]
+	if !ok {
+		return "", errors.New("uri not found in ratelimit redis secret")
+	}
+
+	return string(host), nil
+}
+
+// requestEndpoint sends a simple GET request to the endpoint and returns an
+// error if the response is not OK
+func requestEndpoint(endpoint string) error {
+	c := &http.Client{}
+	res, err := c.Get(endpoint)
+	if err != nil {
+		return err
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("unexpected status code. Expected %d got %d", http.StatusOK, res.StatusCode)
+	}
+
+	return nil
+}
+
+type redisCounter struct {
+	client kubernetes.Interface
+	config *rest.Config
+
+	PodName   string
+	Namespace string
+	RedisHost string
+}
+
+// GetCount uses the throwaway Redis pod to obtain the value of the counter
+// in the rate limit Redis instance
+func (r *redisCounter) GetCount() (int, error) {
+	result, err := testcommon.ExecToPodArgs(r.client, r.config,
+		[]string{
+			"/opt/rh/rh-redis32/root/usr/bin/redis-cli",
+			"-c",
+			"-h",
+			r.RedisHost,
+			"-p",
+			"6379",
+			"KEYS",
+			"*",
+		},
+		r.PodName,
+		r.Namespace,
+		"redis",
+	)
+	if err != nil {
+		return 0, err
+	}
+
+	result = strings.TrimSpace(strings.ReplaceAll(result, "liveness-probe", ""))
+
+	if result == "" {
+		return 0, nil
+	}
+
+	count, err := testcommon.ExecToPod(r.client, r.config,
+		fmt.Sprintf("/opt/rh/rh-redis32/root/usr/bin/redis-cli -c -h %s -p 6379 GET %s", r.RedisHost, result),
+		r.PodName,
+		r.Namespace,
+		"redis",
+	)
+	if err != nil {
+		return 0, err
+	}
+	count = strings.TrimSpace(count)
+
+	return strconv.Atoi(count)
```

Comment:
I got an error after running this test against RHOAM installed from master branch:

```
...
  ☁️  Waiting for throwaway Redis container to complete. Current phase: in progress...
  ☁️  Waiting for throwaway Redis container to complete. Current phase: in progress...
 ️ℹ️  Throw away Redis Pod ready
  Redis pod name: throw-away-redis-pod-57f8ddf94f-hsk76
Previous count: 0 | Number of requests: 10 | Current count: 10 | [PASS] 🎉
Previous count: 0 | Number of requests: 14 | Current count: 14 | [PASS] 🎉
strconv.Atoi: parsing "ERR wrong number of arguments for 'get' command": invalid syntax
```

Reply:
After I removed the redis CR manually, I tried to rerun the test, but it failed with
```bash
  ☁️  Waiting for throwaway Redis container to complete. Current phase: in progress...
INFO[0004] Looking for RHMI CR                           ns=redhat-rhoam-operator
  🔑  Found access token: <REDACTED>
unexpected response code: 422
exit status 1
```

Reply:
Hmmm I was getting this kind of errors randomly but after adding the flush call it was working consistently, but seems like it's still occurring. I'll investigate what's causing this.

I think it's failing to re-run because the API was already created. A workaround is changing the base name in the code, so it will just create a new one:
https://github.com/sergioifg94/integreatly-operator/blob/MGDAPI-1261/test/scripts/products/h22-validate-that-rate-limit-service-is-working-as-expected/h22-validate-that-rate-limit-service-is-working-as-expected.go#L82

Reply:
@psturc  I just pushed some changes that fix all these reliability issues. I tested it with 50 iterations a few times and it's working consistently now

Reply:
@psturc I just pushed the changes you suggested about cleaning up the testing API. It will clean up the API after the test, but seems like 3scale might take a few minutes to pick up those changes so I left a note to wait around 5 minutes before retesting.
To know when it's not yet ready to re-test, in the dashboard it would still show the test backend in the dropdown below, but when clicking on it we get 404. After a few minutes that is gone and re-launching the test will work
![image](https://user-images.githubusercontent.com/12643415/110648554-b7094f80-81b0-11eb-9a9f-91f134091abc.png)